File: app/src/main/java/com/eh/recyclerlistsample/NoteAdapter.java
package com.eh.recyclerlistsample;

import android.graphics.Color;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Random;

/**
 * Created by IT SCHOOL on 14.02.2018.
 */
public class NoteAdapter extends
        RecyclerView.Adapter<NoteAdapter.NoteViewHolder> {

    private final ArrayList<NoteModel> notes = new ArrayList<>();

    @Override
    public NoteViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        // описываем как создается NoteViewHolder
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.rv_note_item, parent, false);
        return new NoteViewHolder(view);
    }

    @Override
    public void onBindViewHolder(NoteViewHolder holder, int position) {
        // описываем как заполняется NoteViewHolder
        NoteModel model = notes.get(position);
        holder.title.setText(model.getTitle());
        holder.text.setText(model.getText());
        holder.date.setText(model.getFormattedDate());
        Random rand = new Random();
        int color = Color.argb(255, rand.nextInt(255), rand.nextInt(255), rand.nextInt(255));
        holder.category.setBackgroundColor(color);
    }

    @Override
    public int getItemCount() {
        return notes.size();
    }

    public void updateNotes(ArrayList<NoteModel> notes) {
        this.notes.clear();
        this.notes.addAll(notes);
        notifyDataSetChanged(); // обновляет все видимые элементы
    }

    static class NoteViewHolder extends RecyclerView.ViewHolder {
        // Один элемент списка
        TextView title, text, date;
        View category;

        public NoteViewHolder(View itemView) {
            super(itemView);
            title = (TextView) itemView.findViewById(R.id.tv_title);
            text = (TextView) itemView.findViewById(R.id.tv_text);
            date = (TextView) itemView.findViewById(R.id.tv_date);
            category = itemView.findViewById(R.id.v_category);
        }

    }
}

File: app/src/main/java/com/eh/recyclerlistsample/NoteAdapter.kt
package com.eh.recyclerlistsample

import android.graphics.Color
import com.eh.recyclerlistsample.NoteAdapter.NoteViewHolder
import android.view.ViewGroup
import android.view.LayoutInflater
import android.view.View
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView.Adapter
import com.eh.recyclerlistsample.R.id
import com.eh.recyclerlistsample.R.layout
import java.util.ArrayList
import java.util.Random

class NoteAdapter : Adapter<NoteViewHolder>() {
  private val notes = ArrayList<NoteModel>()
  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): NoteViewHolder {
    // описываем как создается NoteViewHolder
    val view = LayoutInflater.from(parent.context)
      .inflate(layout.rv_note_item, parent, false)
    return NoteViewHolder(view)
  }

  override fun onBindViewHolder(holder: NoteViewHolder, position: Int) {
    // описываем как заполняется NoteViewHolder
    val model = notes[position]
    holder.title.text = model.title
    holder.text.text = model.text
    holder.date.text = model.formattedDate
    val rand = Random()
    val color = Color.argb(255, rand.nextInt(255), rand.nextInt(255), rand.nextInt(255))
    holder.category.setBackgroundColor(color)
  }

  override fun getItemCount(): Int {
    return notes.size
  }

  fun updateNotes(notes: ArrayList<NoteModel>?) {
    this.notes.clear()
    this.notes.addAll(notes!!)
    notifyDataSetChanged() // обновляет все видимые элементы
  }

  class NoteViewHolder(itemView: View) : ViewHolder(itemView) {
    // Один элемент списка
    var title: TextView
    var text: TextView
    var date: TextView
    var category: View

    init {
      title = itemView.findViewById<View>(id.tv_title) as TextView
      text = itemView.findViewById<View>(id.tv_text) as TextView
      date = itemView.findViewById<View>(id.tv_date) as TextView
      category = itemView.findViewById(id.v_category)
    }
  }
}